---
title: "Permutation test exercise"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
library(tidyverse)
library(GenomicRanges)
source("analysis/util/intersect_functions.R")
source("analysis/util/_setup.R")

```

## Purpose: Determine statistical enrichment depletion of each peak set vs. sets of regions.


```{r import}

# Get peaks and promoters

peak_list <- import_peaks("/Shares/rinn_class/data/CLASS_2020/analysis/01_consensus_peaks/results")

gencode_gr <- rtracklayer::import("/Shares/rinn_class/data/genomes/human/gencode/v32/gencode.v32.annotation.gtf")
lncrna_promoters <- get_promoter_regions(gencode_gr, biotype = "lncRNA")
mrna_promoters <- get_promoter_regions(gencode_gr, biotype = "protein_coding")

# rmsk <- import_repeatmasker()
# rmsk_family <- subset_rmsk(rmsk, rep_level = "family")
# names(rmsk_family) <- paste(names(rmsk_family), "family", sep = "_")
# rmsk_class <- subset_rmsk(rmsk, rep_level = "class")
# names(rmsk_class) <- paste(names(rmsk_class), "class", sep = "_")

# Combine both biotypes to one list
region_list <- c("lncrna_promoters" = list(lncrna_promoters), 
                 "mrna_promoters" = list(mrna_promoters))


# Remove non-canonical chromosomes
canonical_chr <- c(paste0("chr", 1:22), "chrM", "chrX", "chrY")

for(i in 1:length(region_list)) {
  region_list[[i]] <- region_list[[i]][which(seqnames(region_list[[i]]) %in% canonical_chr)]
}

for(i in 1:length(peak_list)) {
  tmp_pl <- peak_list[[i]][which(seqnames(peak_list[[i]]) %in% canonical_chr)]
  # Workaround to remove the factors in the seqnames that are not canonical chromosomes.
  tmp_pl <- tmp_pl %>% as.data.frame()
  tmp_pl <- GRanges(seqnames = as.character(tmp_pl$seqnames),
                    ranges = IRanges(start = tmp_pl$start,
                                     end = tmp_pl$end),
                    strand = tmp_pl$strand,
                    name = tmp_pl$name)
  peak_list[[i]] <- tmp_pl
}

# Remove peak files with no consensus peaks
peak_list <- peak_list[sapply(peak_list,length)>0]

```


```{r}

# Import the length of each chromosome
chr_len <- read_csv("/Shares/rinn_class/data/k562_chip/analysis/11_peak_feature_intersect/results/chr_sizes.csv") %>%
  column_to_rownames("chrom")

# Shuffle peaks, maintaining width and chromosome enrichment
shuffle_peaks <- function (peak_gr, genome) {
  chrLens <- genome[names(seqlengths(peak_gr)),"size"]
  nn <- as.vector(seqnames(peak_gr))
  ii <- order(nn)
  w <- width(peak_gr)
  nnt <- table(nn)
  jj <- order(names(nnt))
  nnt <- nnt[jj]
  chrLens <- chrLens[jj]
  ss <- unlist(sapply(1:length(nnt), function(i) sample(chrLens[i], 
                                                        nnt[i])))
  res <- GRanges(seqnames = nn[ii], ranges = IRanges(ss, width = w[ii]), 
                 strand = "*")
  return(res)
}


# This is a wrapper function which will call shuffle_peaks a number of times
# and return each permuted peak set as a list.
make_permutations <- function(region_set, genome, nperms) {
  perms <- list()
  for(i in 1:nperms) {
    shuf <- shuffle_peaks(region_set, genome)
    perms <- c(perms, list(shuf))
  }
  return(perms)
}

```

```{r}

# Call the make permutations function
permuted_peaks <- make_permutations(peak_list[[1]], chr_len, nperms = 1000)

# Determine inflation rate of one peak list
as.numeric(object.size(permuted_peaks))/as.numeric(object.size(peak_list[1:4]))

# View(p.adjust)

set.seed(12343)
peak_list_nulls <- lapply(peak_list[1:4], make_permutations, genome = chr_len, nperms=1000)

```

```{r}

# Calculate the number of overlaps for each of our region sets (mRNA and lncRNA)

resdf <- tibble(region = character(),
                tf = character(),
                observed = integer(),
                permuted = character())

for(i in 1:length(region_list)) {
  region_set_name <- names(region_list)[i]
  region_set <- region_list[[i]]

  res <- lapply(peak_list, function(peaks) { 
    length(which(countOverlaps(region_set, peaks) > 0)) 
  }) 

  res <- res %>%
    as.data.frame() %>%
    pivot_longer(everything(), names_to = "tf", values_to = "observed")
  
  perm_ovs <- lapply(peak_list_nulls, function(peak_nulls) {
    permuted <- c()
    for(j in 1:length(peak_nulls)) {
      if(class(peak_nulls[[j]]) != "GRanges") { stop(paste0("Not a Granges object")) }
      permuted <- c(permuted, length(which(countOverlaps(region_set, peak_nulls[[j]]) > 0)))
    }
    permuted <- paste(permuted, collapse = ";")
    return(permuted)
  })
  
  perm_ovsdf <- data.frame("tf" = names(unlist(perm_ovs)), 
                           "permuted" = unlist(perm_ovs))

  res <- merge(res, perm_ovsdf)
  res$region <- region_set_name
  res <- res %>% dplyr::select(region, everything())
  resdf <- bind_rows(resdf, res)
}
```

```{r}
# Calculate a zscore and a p-value based on that null distribution.

resdf_long <- resdf %>%
  separate_rows(permuted, sep = ";") %>%
  mutate(permuted = as.numeric(permuted))

resdf_summary <- resdf_long %>% 
  group_by(region, tf) %>%
  summarize(observed = unique(observed),
            mean_permuted = mean(permuted),
            diff = observed - mean_permuted,
            alternative = ifelse(unique(observed) < mean(permuted), "less", "greater"),
            zscore = round((unique(observed) - mean(permuted))/stats::sd(permuted), 4),
            pval = ifelse(unique(alternative) == "less", (sum(unique(observed) >= permuted, na.rm = TRUE) + 1)/(1000 + 1),
                          (sum(unique(observed) <= permuted, na.rm = TRUE) + 1)/(1000 + 1)))

hist(resdf_summary$pval)

resdf_summary$padj <- p.adjust(resdf_summary$pval, method = "bonferroni")

filter(resdf_summary, padj < 0.05)

```

### Plotting

Visualize some of the null distributions and how the number of observed peaks compares

```{r}

single_result <- resdf_long %>% filter(tf == "AFF1", region == "lncrna_promoters")
g <- ggplot(single_result, aes(x = permuted))
g + geom_bar() + 
  geom_vline(xintercept = unique(single_result$observed), lty = 2) +
  theme_paperwhite() + 
  ggtitle("Permuted overlaps vs. observed",
    subtitle = paste0("TF: ", unique(single_result$tf), " -- Region: ", unique(single_result$region))) + 
  xlab("Number of overlaps") + 
  ylab("Count") 
# ggsave(paste0("figures/permuted_vs_observed_",unique(single_result$tf), "_", unique(single_result$region), ".pdf"))

```

```{r}

# Make a heatmap of results
g <- ggplot(resdf_summary, aes(x = tf, y = region, fill = zscore))
g + geom_raster() + scale_fill_gradient2() + coord_flip() + 
  theme(axis.text.x = element_text(angle = 90L, hjust = 1L, vjust = 0.5))

resdf_matrix <- resdf_summary %>%
  dplyr::select(tf, region, zscore) %>%
  pivot_wider(names_from = tf, values_from = zscore) %>%
  column_to_rownames("region") %>%
  as.matrix()

tf_clust <- hclust(dist(t(resdf_matrix)))
plot(tf_clust)

resdf_summary$tf <- factor(resdf_summary$tf, tf_clust$labels[tf_clust$order])

g <- ggplot(resdf_summary, aes(x = tf, y = region, fill = zscore))
g + geom_raster() + scale_fill_gradient2() + coord_flip() + 
  theme(axis.text.x = element_text(angle = 90L, hjust = 1L, vjust = 0.5))

# ggsave("figures/lncrna_mrna_promoters_vs_tf_peaks.pdf", height = 5, width = 5)

```





